### PR TITLE
feat(lang): Support julia

### LINKED ISSUE #408: feat: Support Julia language
## Description

Currently, we only support Python and R (experimental). Let's discuss if there's enough interest to support Julia in this issue.

### PR DESCRIPTION
**Pull request by gaocegege**
Close #408 

Signed-off-by: Ce Gao <cegao@tensorchord.ai>

### CODE REVIEW

File: base-images/julia1.8rc1-ubuntu20.04.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+RUN set -x && \
+    UNAME_M="$(uname -m)" && \
+    if [ "${UNAME_M}" = "x86_64" ]; then \
+        MINICONDA_URL="https://repo.anaconda.com/miniconda/Miniconda3-${CONDA_VERSION}-Linux-x86_64.sh"; \
+        SHA256SUM="4ee9c3aa53329cd7a63b49877c0babb49b19b7e5af29807b793a76bdb1d362b4"; \
+    elif [ "${UNAME_M}" = "s390x" ]; then \
+        MINICONDA_URL="https://repo.anaconda.com/miniconda/Miniconda3-${CONDA_VERSION}-Linux-s390x.sh"; \
+        SHA256SUM="e5e5e89cdcef9332fe632cd25d318cf71f681eef029a24495c713b18e66a8018"; \
+    elif [ "${UNAME_M}" = "aarch64" ]; then \
```
Comment (terrytangyuan) [inline base-images/julia1.8rc1-ubuntu20.04.Dockerfile:39]: Conda is not needed?

File: base-images/build.sh [hunk 2]
```diff
@@ -15,30 +15,38 @@ fi
...
 docker buildx build \
     --build-arg ENVD_VERSION=${ENVD_VERSION} \
     --build-arg ENVD_SSH_IMAGE=ghcr.io/tensorchord/envd-ssh-from-scratch \
     --build-arg HTTP_PROXY=${HTTP_PROXY} \
     --build-arg HTTPS_PROXY=${HTTPS_PROXY} \
-    -t ${DOCKER_HUB_ORG}/r-base:4.2  \
+    -t ${DOCKER_HUB_ORG}/julia:1.8rc1-ubuntu20.04 \
     --pull --push --platform linux/x86_64,linux/arm64 \
-    -f r4.2.Dockerfile .
-
```
Comment (terrytangyuan) [inline base-images/build.sh:43]: Still WIP?

File: pkg/lang/ir/compile.go [hunk 3]
```diff
@@ -154,6 +161,49 @@ func (g Graph) Compile(uid, gid int) (llb.State, error) {
...
+	if err != nil {
+		return llb.State{}, errors.Wrap(err, "failed to compile shell")
+	}
+	diffShellStage := llb.Diff(builtinSystemStage, shellStage, llb.WithCustomName("install shell"))
+
+	systemStage := llb.Diff(builtinSystemStage, g.compileSystemPackages(builtinSystemStage),
+		llb.WithCustomName("install system packages"))
+
+	// TODO(terrytangyuan): Support RStudio local server
+	juliaStage := llb.Diff(builtinSystemStage,
```
Comment (terrytangyuan) [inline pkg/lang/ir/compile.go:184]: This would need some refactoring to reduce code redundancy.

**Comment by terrytangyuan** [comment]
Exciting! We need to refresh README soon to reflect the new supported languages.

File: base-images/julia1.8rc1-ubuntu20.04.Dockerfile (hunk 1 above)
Comment (gaocegege) [reply base-images/julia1.8rc1-ubuntu20.04.Dockerfile:39]: Yep I think we can remove it. I was thinking if we should support python in julia's image Ref [#same](https://github.com/tensorchord/envd/issues/407)

File: pkg/lang/ir/compile.go (hunk 3 above)
Comment (gaocegege) [reply pkg/lang/ir/compile.go:184]: Yep I think we need a base graph and extend it with different sub-struct like PythonGraph and JuliaGraph.

**Comment by gaocegege** [comment]
> You may be interested in the Julia - Python interop configuration we setup over on conda-forge: https://github.com/conda-forge/julia-feedstock/blob/main/recipe/scripts/activate.sh

Thanks! I will have a look! :tada:

**Comment by gaocegege** [comment]
PTAL @terrytangyuan @VoVAllen

File: pkg/lang/ir/compile.go [hunk 4]
```diff
@@ -154,6 +161,49 @@ func (g Graph) Compile(uid, gid int) (llb.State, error) {
...
+	shellStage, err := g.compileShell(builtinSystemStage)
+	if err != nil {
+		return llb.State{}, errors.Wrap(err, "failed to compile shell")
+	}
+	diffShellStage := llb.Diff(builtinSystemStage, shellStage, llb.WithCustomName("install shell"))
+
+	systemStage := llb.Diff(builtinSystemStage, g.compileSystemPackages(builtinSystemStage),
+		llb.WithCustomName("install system packages"))
+
+	// TODO(terrytangyuan): Support RStudio local server
```
Comment (terrytangyuan) [inline pkg/lang/ir/compile.go:183]: Remove this?

**Review by terrytangyuan — APPROVED** [review]
Some minor comments. Great work!

File: pkg/lang/ir/types.go [hunk 5]
```diff
@@ -35,13 +35,15 @@ type Graph struct {
 
 	UbuntuAPTSource   *string
 	CRANMirrorURL     *string
+	JuliaPkgServer    *string
```
[the remark this answers is not in the document]
Comment (gaocegege) [reply pkg/lang/ir/types.go:38]: SGTM!

**Comment by gaocegege** [comment]
> You may be interested in the Julia - Python interop configuration we setup over on conda-forge: https://github.com/conda-forge/julia-feedstock/blob/main/recipe/scripts/activate.sh

@mkitti Sorry I am not quite familiar with Julia, could you please explain more about the script? Does it use conda to manage julia packages?

File: pkg/lang/frontend/starlark/config/config.go [hunk 6]
```diff
@@ -140,6 +142,27 @@ func ruleFuncCRANMirror(thread *starlark.Thread, _ *starlark.Builtin,
...
+
+	if err := starlark.UnpackArgs(ruleJuliaPkgServer, args, kwargs,
+		"url?", &url); err != nil {
+		return nil, err
+	}
+
+	urlStr := ""
+	if url != starlark.String("") {
+		urlStr = url.GoString()
+	}
```
Comment (VoVAllen) [inline pkg/lang/frontend/starlark/config/config.go:157]: Can we skip the if statement here? Does `urlStr = url.GoString()` work if `url` is `starlark.String("")`?

File: pkg/progress/progressui/display.go [hunk 7]
```diff
@@ -805,14 +793,14 @@ func (disp *display) print(d displayInfo, width, height int, all bool) {
...
-		statusStr = "✅ (finished)"
+		statusStr = "FINISHED"
 	}
 
 	fmt.Fprint(disp.c, aec.Hide)
 	defer fmt.Fprint(disp.c, aec.Show)
 
-	out := fmt.Sprintf("[+] 🐋 %s %.1fs (%d/%d) %s", disp.phase, time.Since(d.startTime).Seconds(), d.countCompleted, d.countTotal, statusStr)
-	// out = align(out, "", width)
+	out := fmt.Sprintf("[+] %s %.1fs (%d/%d) %s", disp.phase, time.Since(d.startTime).Seconds(), d.countCompleted, d.countTotal, statusStr)
```
Comment (VoVAllen) [inline pkg/progress/progressui/display.go:802]: Does this fix the indent problem?

**Review by VoVAllen — APPROVED** [review]
LGTM!

File: pkg/progress/progressui/display.go (hunk 7 above)
Comment (gaocegege) [reply pkg/progress/progressui/display.go:802]: Yep.

Comment (gaocegege) [reply pkg/progress/progressui/display.go:802]: But the main reason is to fix one flaky test caused by the function.

**Pull request merged**